### ISSUE TITLE
run_regularly() not working for subgroups

### ISSUE THREAD

**Issue by AI-pha** [brian-team/brian2#922]
Hello,

In #766 (May 23, 2017) run_regulary on subgroups was called a "legitimate use case", so I think this is expected to work. However, in the following minimal example, the state variable is not updated by run_regularly. (It does not raise an error and I did not find anything in the documentation concerning run_regulary for subgroups).

Here is the code that reproduces the issue (v just stays at 1, instead of rising by 1 every 10 ms):
```python
from brian2 import NeuronGroup, StateMonitor, run, ms, figure, plot, show

neurons = NeuronGroup(10, model="v:1")
neurons[0:5].v = 1
neurons[0:5].run_regularly("v += 1", dt=10 * ms)

statemon = StateMonitor(neurons, variables=["v"], record=True)

run(100 * ms)

print(neurons[0:5].v)

fig = figure()
plot(statemon.t / ms, statemon.v.T)
show()
```

Thanks a lot for having a look at this issue.

**Comment by mstimberg** [comment]
Hi, many thanks for the report, this in indeed a bug, your code should work as you intended. I had a quick look at it and I know what's going wrong, it will be quite easy to fix.
I'll give some more details on what is happening later, but fortunately there's a quite simple workaround: store the `StateUpdater` object that `run_regularly` returns in a variable to make the `run` function aware of it (if you explicitly construct a `Network` object, add the returned object via `Network.add`). I.e., instead of:
```Python
neurons[0:5].run_regularly("v += 1", dt=10 * ms)
```
use:
```Python
v_updater = neurons[0:5].run_regularly("v += 1", dt=10 * ms)
```

**Comment by AI-pha** [comment]
Thanks for the very quick reply.

When I use:
```python
v_updater = neurons[0:5].run_regularly("v += 1", dt=10 * ms) 
```
I get
`An error occurred when preparing an object. ReferenceError: weakly-referenced object no longer exists
`
This also happens, when I add the 'v_updater' to a Network.


It does however work fine (both for magic and Network), when a subgroup is explicitly created like that:
```python
subgr = neurons[0:5]
v_updater = subgr.run_regularly("v += 1", dt=10 * ms) 
```

It also works without the 'v_updater', if the subgroup is explicitly created and manually added to a Network (which is anyway my preferred way to do it).

**Comment by mstimberg** [comment]
You are right, sorry for the half-working workaround.

To give some background what is going on: when Brian runs a simulation, it simulates a bunch of `CodeObject`s that are responsible for doing the actual computational work. E.g. a `NeuronGroup` typically creates three such objects, one for the "state update" (integration of the differential equations), one for the threshold check and one for the application of the reset statement(s). The user only adds the main object (e.g. the `NeuronGroup`) to a network (or lets the magic system do it automatically), the code objects are automatically included because they are referenced in the `contained_objects` attribute of the main object. Now, normally there is no need to explicitly add a subgroup to a network, since it does not need to compute anything during a run. When the subgroup declares a `run_regularly` operation, this changes, because this `run_regularly` operation is only stored in the `contained_objects` attribute of the subgroup. 

The fix in PR #924 now makes `run_regularly` add the subgroup itself to `contained_objects` of its parent. The subgroup and the referenced `run_regularly` object will therefore be automatically included in the network when the parent group is included.

**Comment by AI-pha** [comment]
Great, thanks a lot also for the background.
